### PR TITLE
Convert Win32 project to .Net Standard

### PR DESCRIPTION
**Pull request by RobGallo**
#### Describe the change

- Have new project use .Net Standard equivalent NuGet packages for System.Drawing and Microsoft.Win32.Registry
- Enable Win32 project unit testing without Fakes
	- Add Microsoft.Win32.Registry (.Net Standard compatible) to SystemAbstractions
	- Make Win32Helper an object
	- Make Win32Helper use SystemAbstractions to access the registry
	- 	Replace Fakes with mocks in Win32Tests project

Note: Action item to organize SystemAbstractions in a more conventional way

> Note: After the PR has been created, certain checks will be kicked off. All of these checks must pass before a merge.

### CODE REVIEW

File: src/SystemAbstractionsTests/MicrosoftUnitTests.cs [hunk 1]
```diff
@@ -0,0 +1,27 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+using Axe.Windows.SystemAbstractions;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using Microsoft.Win32;
+using Moq;
+
+namespace SystemAbstractionsTests
+{
+[TestClass]
```
Comment (DaveTryon) [inline src/SystemAbstractionsTests/MicrosoftUnitTests.cs]: Indents
```suggestion
    [TestClass]
```

File: src/SystemAbstractionsTests/MicrosoftUnitTests.cs [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    public class MicrosoftUnitTests
+    {
+        const string WindowsVersionRegKey = @"HKEY_LOCAL_MACHINE\Software\Microsoft\Windows NT\CurrentVersion";
```
Comment (DaveTryon) [inline src/SystemAbstractionsTests/MicrosoftUnitTests.cs]: >const [](start = 8, length = 5)

Seems odd to have a class-level const that only gets used in 1 method. I'd suggest creating 2 consts (WindowsVersionRegKey and CurrentVersion) inside the test method, then passing the const values to both the "direct" and "indirect" methods being called

**Comment by DaveTryon** [comment]
> // Copyright (c) Microsoft. All rights reserved.

InternalsVisibleTo.cs is a new pattern. Seems like it ought to be in the documentation if it's the way we want to handle internals with .NET standard projects

---
Refers to: src/Win32/InternalsVisibleTo.cs:1 in 5b372e2. [](commit_id = 5b372e2daace8bff927bb417f130a28dac655a82, deletion_comment = False)

File: src/Win32Tests/packages.config [hunk 3]
```diff
@@ -1,5 +1,9 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <packages>
+  <package id="Castle.Core" version="4.4.0" targetFramework="net471" />
+  <package id="Moq" version="4.12.0" targetFramework="net471" />
```
Comment (DaveTryon) [inline src/Win32Tests/packages.config]: >12 [](start = 31, length = 2)

The rest of Axe.Windows is using Moq 4.11. Can we use that version here, as well? (Happy to update the version, but that feels like a separate PR)

File: src/Win32Tests/Win32ApisTests.cs [hunk 4]
```diff
@@ -1,267 +1,223 @@
...
     [TestClass]
     public class Win32ApisUnitTests
     {
+        private readonly Mock<IMicrosoftWin32Registry> _registryMock;
+        private readonly Win32Helper _win32Helper;
+
+        public Win32ApisUnitTests()
+        {
+            _registryMock = new Mock<IMicrosoftWin32Registry>(MockBehavior.Strict);
+                _win32Helper = new Win32Helper(_registryMock.Object);
```
Comment (DaveTryon) [inline src/Win32Tests/Win32ApisTests.cs]: Indents
```suggestion
            _win32Helper = new Win32Helper(_registryMock.Object);
```

File: src/Win32Tests/Win32ApisTests.cs [hunk 5]
```diff
@@ -1,267 +1,223 @@
...
+        }
+
+        [TestInitialize]
+        public void TestInitialize()
+        {
+            _registryMock.Reset();
```
Comment (DaveTryon) [inline src/Win32Tests/Win32ApisTests.cs]: >_registryMock [](start = 12, length = 13)

This pattern requires test to run sequentially. That's probably OK unless someone uses ReSharper's test engine (which has the option to run them in parallel)

File: src/Win32Tests/Win32ApisTests.cs [hunk 6]
```diff
@@ -1,267 +1,223 @@
...
+        private readonly Win32Helper _win32Helper;
+
+        public Win32ApisUnitTests()
+        {
+            _registryMock = new Mock<IMicrosoftWin32Registry>(MockBehavior.Strict);
+                _win32Helper = new Win32Helper(_registryMock.Object);
+        }
+
+        [TestInitialize]
+        public void TestInitialize()
```
Comment (DaveTryon) [inline src/Win32Tests/Win32ApisTests.cs]: >TestInitialize [](start = 20, length = 14)

How about adding a TestCleanup method that calls _registryMock.VerifyAll()  ?

File: src/Win32/Win32Helper.cs [hunk 7]
```diff
@@ -9,20 +9,33 @@ namespace Axe.Windows.Win32
     /// <summary>
     /// NativeMethods partial class to hold all Win32 related helper methods.
```
Comment (DaveTryon) [inline src/Win32/Win32Helper.cs]: >NativeMethods [](start = 8, length = 13)

Comment refers to old class

**Comment by DaveTryon** [comment]
>     <TargetFrameworkVersion>v4.7.1</TargetFrameworkVersion>

Win32Tests is still using .NET Framework. Is it a separate PR to convert the test project?

---
Refers to: src/Win32Tests/Win32Tests.csproj:12 in 5b372e2. [](commit_id = 5b372e2daace8bff927bb417f130a28dac655a82, deletion_comment = False)

File: src/Win32/Win32Helper.cs [hunk 8]
```diff
@@ -55,21 +68,23 @@ internal static void GetDpi(Point point, DpiType dpiType, out uint dpiX, out uin
         /// Get the current Windows version info from HKLM
         /// </summary>
         /// <returns></returns>
-        private static string GetCurrentWindowsVersion()
+        private string GetCurrentWindowsVersion()
         {
-            return (string)Registry.GetValue(WindowsVersionRegKey, "CurrentVersion", "");
+            var retVal = (string)_registry.GetValue(WindowsVersionRegKey, "CurrentVersion", "");
+            System.Diagnostics.Debug.WriteLine($"retVal: {retVal}");
```
Comment (DaveTryon) [inline src/Win32/Win32Helper.cs]: >WriteLine [](start = 37, length = 9)

Is this line necessary?

File: src/SystemAbstractionsTests/MicrosoftUnitTests.cs [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+
+        private readonly IMicrosoft _microsoft = MicrosoftFactory.CreateMicrosoft();
+
+        [TestMethod]
+        [Timeout(1000)]
+        public void RegistryGetValue_Matches()
+        {
+            var expectedValue = Registry.GetValue(WindowsVersionRegKey, "CurrentVersion", "");
```
Comment (DaveTryon) [inline src/SystemAbstractionsTests/MicrosoftUnitTests.cs]: >"" [](start = 90, length = 2)

I find that string.Empty is more expressive and more useful when trying to search for code. YMMV.

File: src/Desktop/Utility/ExtensionMethods.cs [hunk 10]
```diff
@@ -183,7 +183,8 @@ public static double GetDPI(this Rectangle rc)
         /// <returns></returns>
         public static double GetDPI(int left, int top)
         {
-            NativeMethods.GetDpi(new System.Drawing.Point(left, top), DpiType.Effective, out uint dpiX, out uint dpiY);
+            var win32Helper = new Win32Helper();
```
Comment (DaveTryon) [inline src/Desktop/Utility/ExtensionMethods.cs]: >Win32Helper [](start = 34, length = 11)

This object will get created and destroyed many times during the process. We also create one in ExtensionMethods.cs. That seems a little bit odd for a class that will be functionally identical in each and every instance

File: src/SystemAbstractions/Abstractions/IMicrosoft.cs [hunk 11]
```diff
@@ -0,0 +1,11 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+using System;
+
+namespace Axe.Windows.SystemAbstractions
+{
+    public interface IMicrosoft
```
Comment (DaveTryon) [inline src/SystemAbstractions/Abstractions/IMicrosoft.cs]: >IMicrosoft [](start = 21, length = 10)

With this change, can we also remove the Shim for Microsoft.Win32.Registry from mscorlib.fakes?

File: src/SystemAbstractionsTests/MicrosoftUnitTests.cs (hunk 2 above)
Comment (RobGallo) [reply src/SystemAbstractionsTests/MicrosoftUnitTests.cs]: It is used in two methods.

File: src/Win32Tests/packages.config (hunk 3 above)
Comment (RobGallo) [reply src/Win32Tests/packages.config]: Updated

File: src/Win32Tests/Win32ApisTests.cs (hunk 5 above)
Comment (RobGallo) [reply src/Win32Tests/Win32ApisTests.cs]: Noted. Given the existence of the TestInitializeAttribute and it's serial implications though, I feel comfortable with this.

File: src/Win32Tests/Win32ApisTests.cs (hunk 6 above)
Comment (RobGallo) [reply src/Win32Tests/Win32ApisTests.cs]: Done.

File: src/SystemAbstractions/Abstractions/IMicrosoft.cs (hunk 11 above)
Comment (RobGallo) [reply src/SystemAbstractions/Abstractions/IMicrosoft.cs]: Done.

File: src/SystemAbstractionsTests/MicrosoftUnitTests.cs (hunk 9 above)
Comment (RobGallo) [reply src/SystemAbstractionsTests/MicrosoftUnitTests.cs]: Updated.

File: src/SystemAbstractionsTests/MicrosoftUnitTests.cs (hunk 2 above)
Comment (RobGallo) [reply src/SystemAbstractionsTests/MicrosoftUnitTests.cs]: Done.

File: src/Win32/Win32Helper.cs (hunk 7 above)
Comment (RobGallo) [reply src/Win32/Win32Helper.cs]: Fixed.

File: src/Win32/Win32Helper.cs (hunk 8 above)
Comment (RobGallo) [reply src/Win32/Win32Helper.cs]: Removed.

**Pull request merged**